tests/nn/data_parallel/test_fsdp_fwd_multiple_times.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with forward pass with multiple calls of a module, like a loop. """
+
+import tempfile
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+from torch.nn import Linear, Module
+from torch.optim import Adam
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.nn.data_parallel.fully_sharded_data_parallel import TrainingState
+from fairscale.nn.misc import checkpoint_wrapper
+from fairscale.utils.testing import dist_init, skip_if_single_gpu, teardown, torch_version
+
+
+class MyModel(Module):
+    """Model that forward multiple times through the same module."""
+
+    def __init__(self, params_in_root):
+        super().__init__()
+        self.params_in_root = params_in_root
+        self.child1 = Linear(3, 3, bias=False)
+        self.child2 = Linear(3, 3, bias=False)
+        if self.params_in_root:
+            self.layer_in_root_fsdp = Linear(3, 3, bias=False)
+
+    def forward(self, x):
+        # Same layers, once.
+        x = self.child1(x)
+        x = self.child2(x)
+        # Twice.
+        x = self.child1(x)
+        x = self.child2(x)
+        # Optionally, params in root FSDP.
+        if self.params_in_root:
+            x = self.layer_in_root_fsdp(x)
+        return x
+
+
+def _test_func(rank, world_size, model, fsdp_config, checkpoint, tempfile_name, unused):
+    result = dist_init(rank, world_size, tempfile_name, unused)
+    assert result, "Dist init failed"
+
+    model.cuda()
+    assert isinstance(fsdp_config, dict), str(fsdp_config)
+    assert checkpoint in [True, False], checkpoint
+
+    # Inner FSDPs.
+    model.child1 = FSDP(model.child1, **fsdp_config)
+    model.child2 = FSDP(model.child2, **fsdp_config)
+    # Optionally checkpoint them.
+    if checkpoint:
+        model.child1 = checkpoint_wrapper(model.child1)
+        model.child2 = checkpoint_wrapper(model.child2)
```

Comment:
I didn't get a chance to review/dig into this deeply yet, but leaving a small comment. I think checkpointing needs to be inside the FSDP, otherwise I think it doesn't work. Also, theoretically it's faster to have the checkpoint inside the FSDP, otherwise it would re-gather parameters between the fwd/bwd inside the checkpointed backward

Reply:
interesting! Let me try it. I remember in vissl, we can only have checkpoint outside of fsdp, otherwise, there was an error. Let me gather more info.